### PR TITLE
--public-params flag

### PR DESCRIPTION
**Pull request by AndrewNolte**
This flag makes all parameters public

### CODE REVIEW

**Comment by wsnyder** [comment]
Can you please describe the use case for this?

**Comment by AndrewNolte** [comment]
We would like more runtime introspection on modules without a performance hit, and I believe because params are static, this should only slightly increase build time while not affecting run time. It will also enable us to remove a lot of public on/off in our code. I believe this was the default behavior some years ago.

**Comment by wsnyder** [comment]
Ok, please update docs/guide/exe_verilator.rst to describe the new flag, and bin/verilator to summarize it.

**Comment by AndrewNolte** [comment]
Sorry for the delay, just updated. Could we get this landed first? I think they're going to have merge conflicts https://github.com/verilator/verilator/pull/3952

File: src/V3LinkParse.cpp [hunk 1]
```diff
@@ -258,6 +258,14 @@ class LinkParseVisitor final : public VNVisitor {
             }
         }
 
+        if (v3Global.opt.publicParams()) {
+            switch (nodep->varType()) {
+            case VVarType::GPARAM:  // FALLTHRU
+            case VVarType::LPARAM: nodep->sigUserRWPublic(true); break;
+            default: break;
+            }
+        }
```
Comment (wsnyder) [inline src/V3LinkParse.cpp:267]:
```suggestion
        if (v3Global.opt.publicParams() && nodep->isParam()) nodep->sigUserRWPublic(true);
```

File: src/V3Options.h [hunk 2]
```diff
@@ -263,6 +263,7 @@ class V3Options final {
     bool m_protectIds = false;      // main switch: --protect-ids
     bool m_public = false;          // main switch: --public
     bool m_publicFlatRW = false;    // main switch: --public-flat-rw
+    bool m_public_params = false;   // main switch: --public-params
```
Comment (wsnyder) [inline src/V3Options.h:266]:
```suggestion
    bool m_publicParams = false;    // main switch: --public-params
```

And fix related.

**Comment by wsnyder** [comment]
Thanks for your work on this.

**Pull request merged**